app/controllers/application_controller.rb

```diff
@@ -2,9 +2,140 @@
 
 class ApplicationController < Sinatra::Base
 
-  configure do
-    set :public_folder, 'public'
-    set :views, 'app/views'
-  end
+	configure do
+		set :public_folder, 'public'
+		set :views, 'app/views'
+		enable :sessions
+		set :session_secret, "password_security"
+	end
+
+	get '/' do 
+		erb :index
+	end
+
+	get '/tweets' do 
+		if logged_in?
+			erb :'tweets/tweets'
+		else
+			redirect '/login'
+		end
+	end
+
+	get '/signup' do 
+		if self.logged_in?
+			redirect '/tweets'
+		else
+			erb :'users/create_user'
+		end
+	end
+
+	get '/login' do
+		if logged_in?
+			redirect '/tweets'
+		else
+			erb :'users/login'
+		end
+	end
+
+	get '/logout' do 
+		if logged_in?
+			session.clear
+			redirect '/login'
+		else
+			redirect '/'
+		end
+	end
+
+	get '/users/:slug' do 
+		@user = User.find_by_slug(params[:slug])
+		erb :'users/user_tweets'
+	end
+
+
+	post '/signup' do 
+		@user = User.new
+		@user.username = params[:username]
+		@user.email = params[:email]
+		@user.password = params[:password]
+		if @user.save
+			session[:id] = @user.id
+			redirect '/tweets'
+		else
+			redirect '/signup'
+		end
+	end
+
+	post '/login' do
+		@user = User.find_by(username: params[:username])
```

Comment:
The thing is, if you want to prevent any problems, you would have to prevent the creation of usernames that resolve to the same slug.  E.g. 'The Great' and 'The-Great' would both resolve to the same slug 'the-great' while passing a validation based on username.  To use ActiveRecord validation, the only solution I can think of is to make the slug a database column, which, of course increases the size of the database.